### PR TITLE
Velocity_Gradient optimizer added in NeuralNetLearner in learning.py

### PR DESCRIPTION
**Pull request by abbas5253**
Velocity_Gradient optimizer added in NeuralNet in learning.py which allows choosing a larger value of learning_rate for gradient descent.
I have added an extra argument "momentum" in NeuralNetlearner function in learning.py which is set to False in default
I have also added a test for the Velocity_Gradient in the 'test_learning.py' file

### CODE REVIEW

File: learning.py [hunk 1]
```diff
@@ -655,7 +655,7 @@ def predict(example):
 
 
 def NeuralNetLearner(dataset, hidden_layer_sizes=[3],
-                     learning_rate=0.01, epochs=100, activation = sigmoid):
+                     learning_rate=0.01, epochs=100, activation = sigmoid,momentum=False):
```
Comment (MrDupin) [inline learning.py]: To comply with the overall repo style, add a space after commas.

File: learning.py [hunk 2]
```diff
@@ -668,7 +668,7 @@ def NeuralNetLearner(dataset, hidden_layer_sizes=[3],
     # construct a network
     raw_net = network(i_units, hidden_layer_sizes, o_units, activation)
     learned_net = BackPropagationLearner(dataset, raw_net,
-                                         learning_rate, epochs, activation)
+                                         learning_rate, epochs, activation,momentum=momentum)
```
Comment (MrDupin) [inline learning.py]: Here as well.

File: learning.py [hunk 3]
```diff
@@ -697,7 +697,7 @@ def random_weights(min_value, max_value, num_weights):
     return [random.uniform(min_value, max_value) for _ in range(num_weights)]
 
 
-def BackPropagationLearner(dataset, net, learning_rate, epochs, activation=sigmoid):
+def BackPropagationLearner(dataset, net, learning_rate, epochs, activation=sigmoid,momentum=False,beta=0.903):
```
Comment (MrDupin) [inline learning.py]: Here as well.

File: learning.py [hunk 4]
```diff
@@ -739,7 +739,8 @@ def BackPropagationLearner(dataset, net, learning_rate, epochs, activation=sigmo
 
             # Initialize delta
             delta = [[] for _ in range(n_layers)]
-
+            if momentum == True:
+            	v_dw = [[0 for i in range(len(_))] for _ in net]
```
Comment (MrDupin) [inline learning.py]: I think the indentation should be longer, although it may be a glitch in my browser.

File: learning.py [hunk 5]
```diff
@@ -784,15 +785,44 @@ def BackPropagationLearner(dataset, net, learning_rate, epochs, activation=sigmo
...
+
+            #  Update weights without velocity gradient
+            if momentum == False:
+                for i in range(1, n_layers):
+                    layer = net[i]
+                    inc = [node.value for node in net[i-1]]
+                    units = len(layer)
+                    for j in range(units):
+                        layer[j].weights = vector_add(layer[j].weights,
+                                                    scalar_vector_product(
```
Comment (MrDupin) [inline learning.py]: Again, may be a glitch on how GitHub presents the code, but you should use proper formatting in multi-line code. Like this:

```
function(something,
         something_else(
             other_thing
         )
)
```

Alternatively:

```
function(something,
         something_else(
             other_thing
         ))
```

File: learning.py [hunk 6]
```diff
@@ -784,15 +785,44 @@ def BackPropagationLearner(dataset, net, learning_rate, epochs, activation=sigmo
...
+
+            # Update weights with velocity gradient
+            else:
+                for i in range(1, n_layers):
+                    layer = net[i]
+                    inc = [node.value for node in net[i-1]]
+                    units = len(layer)
+                    for j in range(units):
+                        layer[j].weights = vector_add(layer[j].weights,
+                                                    scalar_vector_product(
```
Comment (MrDupin) [inline learning.py]: Here again.

**Review by antmarakis — COMMENTED** [review]
I cannot comment on the actual code since I am not familiar with this method, but I noticed some stylistic issues.

File: learning.py (hunk 6 above)
Comment (abbas5253) [reply learning.py]: multi-line code style modified

File: learning.py (hunk 5 above)
Comment (abbas5253) [reply learning.py]: multi-line code style modified

File: learning.py (hunk 1 above)
Comment (abbas5253) [reply learning.py]: spaces added

File: learning.py (hunk 2 above)
Comment (abbas5253) [reply learning.py]: space added

File: learning.py (hunk 3 above)
Comment (abbas5253) [reply learning.py]: added

File: learning.py [hunk 7]
```diff
@@ -655,7 +655,7 @@ def predict(example):
 
 
 def NeuralNetLearner(dataset, hidden_layer_sizes=[3],
-                     learning_rate=0.01, epochs=100, activation = sigmoid,momentum=False):
+                     learning_rate=0.01, epochs=100, activation = sigmoid, momentum=False):
```
Comment (MrDupin) [inline learning.py]: Not your commit, but can you also remove spaces from around the equal sign in 'activation = sigmoid'?

Comment (abbas5253) [reply learning.py]: ya sure!

**Comment by dmeoli** [comment]
Thanks @abbas5253. This adds a beyond-the-book optimizer to the 3e NeuralNetLearner; the maintained NN/optimizer code (SGD, adam) lives in deep_learning4e.py. Closing as out of scope for learning.py; thanks for contributing!
